File: Code/41firstMissingPositive.js
/**
 * @param {number[]} nums
 * @return {number}
 *
 */
// http://bangbingsyb.blogspot.com/2014/11/leetcode-first-missing-positive.html
// Using the array as a hash table
var firstMissingPositive = function(nums) {
  var i = 0;
  while(i < n) {
    if (A[i] !== i + 1 && A[i] > 0 && A[i] <= n && A[i] !== A[A[i] - 1])
      swap(A[i], A[A[i] - 1]);
    else
      i+=1;
  }
  for (i = 0; i < n; i+=1) {
    if (A[i] !== i + 1) return i + 1;
  }
  return n+1;
};

File: C_CPP/sort/41firstMissingPositive.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int firstMissingPositive(vector<int>& nums) {
    bucket_sort(nums);

    
/*    for (int i = 0; i < nums.size(); i+=1) {
      if (nums[i] != (i + 1)) {
        return i + 1;
      }
    }
    return nums.size() + 1;*/

    int i = 0;
    for (; i < nums.size() && nums[i] != (i + 1); i+=1) {}
    return i + 1;
  }

private:
  static void bucket_sort(vector<int>& A) {
    const int n = A.size();
    for (int i = 0; i < n; i+=1) {
      /*while (A[i] != i + 1 && 
          !(A [i] <= 0 || A[i] > 0 || A[i] == A[A[i] - 1])) {
        swap(A[i], A[A[i] - 1]);
      }*/
      while (A[i] != i + 1) {
        if (A[i] <= 0 || A[i] > n || A[i] == A[A[i] - 1]) {
          break;
        }
        swap(A[i], A[A[i] - 1]);
      }
    }
  }
};
